### ISSUE TITLE
Docker 4.27 build fails

### ISSUE THREAD

**Issue by groogiam** [docker/for-win#13889]
### Description

After upgrading to Docker 4.27 my project no longer builds. Docker cannot seem to find files that are on the disk.

https://github.com/docker/buildx/pull/10 ERROR: failed to calculate checksum of ref 7e5e69fe-0a93-4e6f-94a2-6d69155a6ff7::ry4vsmtlgm8vymwaaeuecfhcn: failed to walk /var/lib/docker/tmp/buildkit-mount1368382612/MyProject/docker-build: lstat /var/lib/docker/tmp/buildkit-mount1368382612/MyProject/docker-build: no such file or directory

### Reproduce

Docker build my project.

### Expected behavior

Build should succeed as it did in 4.26.1.

### docker version

```bash
Client:
 Cloud integration: v1.0.35+desktop.10
 Version:           25.0.1
 API version:       1.44
 Go version:        go1.21.6
 Git commit:        29cf629
 Built:             Tue Jan 23 23:10:35 2024
 OS/Arch:           windows/amd64
 Context:           default

Server: Docker Desktop 4.27.0 (135262)
 Engine:
  Version:          25.0.1
  API version:      1.44 (minimum version 1.24)
  Go version:       go1.21.6
  Git commit:       71fa3ab
  Built:            Tue Jan 23 23:09:46 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.27
  GitCommit:        a1496014c916f9e62104b33d1bb5bd03b0858e59
 runc:
  Version:          1.1.11
  GitCommit:        v1.1.11-0-g4bccb38
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Version:    25.0.1
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.12.1-desktop.4
    Path:     C:\Program Files\Docker\cli-plugins\docker-buildx.exe
  compose: Docker Compose (Docker Inc.)
    Version:  v2.24.3-desktop.1
    Path:     C:\Program Files\Docker\cli-plugins\docker-compose.exe
  debug: Get a shell into any image or container. (Docker Inc.)
    Version:  0.0.22
    Path:     C:\Program Files\Docker\cli-plugins\docker-debug.exe
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.0
    Path:     C:\Program Files\Docker\cli-plugins\docker-dev.exe
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.21
    Path:     C:\Program Files\Docker\cli-plugins\docker-extension.exe
  feedback: Provide feedback, right in your terminal! (Docker Inc.)
    Version:  v1.0.4
    Path:     C:\Program Files\Docker\cli-plugins\docker-feedback.exe
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v1.0.0
    Path:     C:\Program Files\Docker\cli-plugins\docker-init.exe
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     C:\Program Files\Docker\cli-plugins\docker-sbom.exe
  scout: Docker Scout (Docker Inc.)
    Version:  v1.3.0
    Path:     C:\Program Files\Docker\cli-plugins\docker-scout.exe

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 2
 Server Version: 25.0.1
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: a1496014c916f9e62104b33d1bb5bd03b0858e59
 runc version: v1.1.11-0-g4bccb38
 init version: de40ad0
 Security Options:
  seccomp
   Profile: unconfined
 Kernel Version: 5.15.133.1-microsoft-standard-WSL2
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 7.759GiB
 Name: docker-desktop
 ID: 7e5e69fe-0a93-4e6f-94a2-6d69155a6ff7
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
WARNING: daemon is not using the default seccomp profile
```

### Diagnostics ID

818ADD6C-31C5-4F14-AE3A-C97FCC082F74/20240128172229

### Additional Info

I have opened a similar issue here https://github.com/docker/buildx/issues/2218

**Comment by groogiam** [comment]
This seems to be related to builds done through docker compose using the Visual Studio tooling. I created an issue in that repository as well and referenced this issue.

**Comment by anttix** [comment]
The issue also affects command line builds where build context is not in the same directory as the `Dockerfile `

```shell
$ docker compose build
failed to solve: changes out of order: "shared/hello.txt" ""
$ docker build . -f foo/Dockerfile
[ Succeeds ]
```

Here's the SSCCE as a unified diff:
```diff
diff --git a/docker-compose.yml b/docker-compose.yml
new file mode 100644
index 0000000..5c23147
--- /dev/null
+++ b/docker-compose.yml
@@ -0,0 +1,5 @@
+services:
+  foo:
+    build:
+      context: .
+      dockerfile: foo/Dockerfile
diff --git a/foo/Dockerfile b/foo/Dockerfile
new file mode 100644
index 0000000..917a55a
--- /dev/null
+++ b/foo/Dockerfile
@@ -0,0 +1,3 @@
+FROM alpine:3.14
+COPY shared/hello.txt /app
+ENTRYPOINT ["sleep", "infinity"]
diff --git a/shared/hello.txt b/shared/hello.txt
new file mode 100644
index 0000000..557db03
--- /dev/null
+++ b/shared/hello.txt
@@ -0,0 +1 @@
+Hello World
```

**Comment by atalantus** [comment]
Same problem/error for me after upgrading to version 4.27.1.

In my case the Dockerfile even is in the same directory but under a different name.

```
$ docker compose build
failed to solve: failed to compute cache key: failed to calculate checksum of ref <ref>: failed to walk /var/lib/docker/tmp/<buildkit>/<dir>: lstat /var/lib/docker/tmp/<buildkit>/<dir>: no such file or directory
$ docker build . -f foo.Dockerfile
[ Succeeds ]
```

**Comment by atalantus** [comment]
Well, I just updated to the newly released version 4.27.2 and at least for me everything works normally again.
